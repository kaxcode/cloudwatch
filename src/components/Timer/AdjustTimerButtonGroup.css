.AdjustTimerButtonGroup__PlusBtn {
  margin-right: 20px;
  margin-left: 20px;
}

.AdjustTimerButtonGroup {
  grid-area: adjustBtns;
  display: grid;
  grid-template-columns: repeat(3, 1fr);
}
