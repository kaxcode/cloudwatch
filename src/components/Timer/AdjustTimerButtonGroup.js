import React from 'react';
import { Button } from 'react-materialize';
import { func } from 'prop-types';

const AdjustTimerButtonGroup = props => {
  return (
    <tr>
      <td>
        <div className="hours-set">
          <Button
            id="hours-increase"
            className="plus-btn blue"
            onClick={props.increaseHours}
          >
            +
          </Button>
          <Button
            id="hours-decrease"
            className="minus-btn blue"
            onClick={props.decreaseHours}
          >
            -
          </Button>
        </div>
      </td>
      <td>
        <div className="minutes-set">
          <Button
            id="minutes-increase"
            className="plus-btn blue"
            onClick={props.increaseMinutes}
          >
            +
          </Button>
          <Button
            id="minutes-decrease"
            className="minus-btn blue"
            onClick={props.decreaseMinutes}
          >
            -
          </Button>
        </div>
      </td>
      <td>
        <div className="seconds-set">
          <Button
            waves="light"
            id="seconds-increase"
            className="plus-btn blue"
            onClick={props.increaseSeconds}
          >
            +
          </Button>
          <Button
            id="seconds-decrease"
            className="minus-btn blue"
            onClick={props.decreaseSeconds}
          >
            -
          </Button>
        </div>
      </td>
    </tr>
  );
};

AdjustTimerButtonGroup.propTypes = {
  increaseHours: func.isRequired,
  decreaseHours: func.isRequired,
  increaseMinutes: func.isRequired,
  decreaseMinutes: func.isRequired,
  increaseSeconds: func.isRequired,
  decreaseSeconds: func.isRequired
};

export default AdjustTimerButtonGroup;
