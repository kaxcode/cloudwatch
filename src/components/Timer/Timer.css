.Timer__Heading,
.Timer__Hours,
.Timer__Minutes,
.Timer__Seconds,
.Timer__HoursCount,
.Timer__MinutesCount,
.Timer__SecondsCount {
  text-align: center;
}

.Timer__Card {
  min-width: 450px;
}

.Timer__HoursCount,
.Timer__MinutesCount,
.Timer__SecondsCount {
  margin: 0;
}

.Timer__StartBtn,
.Timer__PauseBtn,
.Timer__CleartBtn {
  margin-left: 8px;
}
