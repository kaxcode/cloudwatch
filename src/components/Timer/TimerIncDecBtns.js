import React from 'react';
import { Button } from 'react-materialize';
import { func, number, bool } from 'prop-types';

const Timer = props => {
  return (
    <tr>
      <td>
        <div className="hours-set">
          <Button
            id="hours-increase"
            className="plus-btn blue"
            onClick={props.increaseHours}
          >
            +
          </Button>
          <Button
            id="hours-decrease"
            className="minus-btn blue"
            onClick={props.decreaseHours}
          >
            -
          </Button>
        </div>
      </td>
      <td>
        <div className="minutes-set">
          <Button
            id="minutes-increase"
            className="plus-btn blue"
            onClick={props.increaseMinutes}
          >
            +
          </Button>
          <Button
            id="minutes-decrease"
            className="minus-btn blue"
            onClick={props.decreaseMinutes}
          >
            -
          </Button>
        </div>
      </td>
      <td>
        <div className="seconds-set">
          <Button
            waves="light"
            id="seconds-increase"
            className="plus-btn blue"
            onClick={props.increaseSeconds}
          >
            +
          </Button>
          <Button
            id="seconds-decrease"
            className="minus-btn blue"
            onClick={props.decreaseSeconds}
          >
            -
          </Button>
        </div>
      </td>
    </tr>
  );
};

Timer.propTypes = {
  onStart: func.isRequired,
  onPause: func.isRequired,
  onClear: func.isRequired,
  handleDismiss: func.isRequired,
  increaseHours: func.isRequired,
  decreaseHours: func.isRequired,
  increaseMinutes: func.isRequired,
  decreaseMinutes: func.isRequired,
  increaseSeconds: func.isRequired,
  decreaseSeconds: func.isRequired,
  timeRemaining: number,
  startClicked: bool
};

export default Timer;
