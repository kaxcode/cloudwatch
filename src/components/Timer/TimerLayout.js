import React from 'react';
import { Row, Button, Card, Col, Table } from 'react-materialize';
import {
  secondsToHour,
  secondsToMinutes,
  seconds
} from '../../utils/humanizeTimer';
import Alert from '../Alert/Alert';
import PropTypes from 'prop-types';

const TimerLayout = props => {
  return (
    <Row>
      <Alert
        msg="Time is up!"
        onDismiss={props.handleDismiss}
        show={props.startClicked && props.timeRemaining <= 0}
      />
      <Col s={12} m={3} offset="m2 l4">
        <Card className="white parent-container" textClassName="black-text">
          <h3>Timer</h3>
          <Table className="setters">
            <thead>
              <tr>
                <th className="timer-heading">Hours</th>
                <th className="timer-heading">Minutes</th>
                <th className="timer-heading">Seconds</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <div className="hours-set">
                    <Button
                      id="hours-increase"
                      className="plus-btn blue"
                      onClick={props.increaseHours}
                    >
                      +
                    </Button>
                    <Button
                      id="hours-decrease"
                      className="minus-btn blue"
                      onClick={props.decreaseHours}
                    >
                      -
                    </Button>
                  </div>
                </td>
                <td>
                  <div className="minutes-set">
                    <Button
                      id="minutes-increase"
                      className="plus-btn blue"
                      onClick={props.increaseMinutes}
                    >
                      +
                    </Button>
                    <Button
                      id="minutes-decrease"
                      className="minus-btn blue"
                      onClick={props.decreaseMinutes}
                    >
                      -
                    </Button>
                  </div>
                </td>
                <td>
                  <div className="seconds-set">
                    <Button
                      waves="light"
                      id="seconds-increase"
                      className="plus-btn blue"
                      onClick={props.increaseSeconds}
                    >
                      +
                    </Button>
                    <Button
                      id="seconds-decrease"
                      className="minus-btn blue"
                      onClick={props.decreaseSeconds}
                    >
                      -
                    </Button>
                  </div>
                </td>
              </tr>
              <tr>
                <td>
                  <h1 className="timer-count">
                    {secondsToHour(props.timeRemaining)}
                  </h1>
                </td>
                <td>
                  <h1 className="timer-count">
                    {secondsToMinutes(props.timeRemaining)}
                  </h1>
                </td>
                <td>
                  <h1 className="timer-count">
                    {seconds(props.timeRemaining)}
                  </h1>
                </td>
              </tr>
              <tr>
                <td>
                  <Button
                    waves="light"
                    className="green start-btn"
                    onClick={props.startTimer}
                  >
                    Start
                  </Button>
                </td>
                <td>
                  <Button
                    waves="light"
                    className="red pause-btn"
                    onClick={props.pauseTimer}
                  >
                    Pause
                  </Button>
                </td>
                <td>
                  <Button
                    waves="light"
                    className="yellow darken-3 clear-btn"
                    onClick={props.clearTimer}
                  >
                    Clear
                  </Button>
                </td>
              </tr>
            </tbody>
          </Table>
        </Card>
      </Col>
    </Row>
  );
};

TimerLayout.propTypes = {
  startTimer: PropTypes.func.isRequired,
  pauseTimer: PropTypes.func.isRequired,
  clearTimer: PropTypes.func.isRequired,
  handleDismiss: PropTypes.func.isRequired,
  increaseHours: PropTypes.func.isRequired,
  decreaseHours: PropTypes.func.isRequired,
  increaseMinutes: PropTypes.func.isRequired,
  decreaseMinutes: PropTypes.func.isRequired,
  increaseSeconds: PropTypes.func.isRequired,
  decreaseSeconds: PropTypes.func.isRequired,
  timeRemaining: PropTypes.number,
  startClicked: PropTypes.bool
};

export default TimerLayout;
