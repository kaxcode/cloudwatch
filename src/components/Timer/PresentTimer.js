import React from 'react';
import {
  secondsToHour,
  secondsToMinutes,
  seconds
} from '../../utils/humanizeTimer';
import Alert from '../Alert/Alert';
import { func, number, bool, object } from 'prop-types';
import MessageBoard from '../MessageBoard/MessageBoard.js';
import Nav from '../Nav/Nav.js';
import './Timer.css';

class PresentTimer extends React.Component {
  componentDidMount() {
    this.forceUpdateInterval = setInterval(() => this.forceUpdate(), 50);
  }
  render() {
    return (
      <main className="Timer">
        <Alert
          onDismiss={this.props.handleDismiss}
          show={this.props.startClicked && this.props.timeRemaining <= 0}
        />
        <Nav />
        <div className="Timer__Container">
          <div className="Timer__Counter">
            <div className="Timer__HoursCount">
              {secondsToHour(localStorage.getItem('timeRemaining'))}
            </div>
            <div className="Timer__Delimiter">:</div>
            <div className="Timer__MinutesCount">
              {secondsToMinutes(localStorage.getItem('timeRemaining'))}
            </div>
            <div className="Timer__Delimiter">:</div>
            <div className="Timer__SecondsCount">
              {seconds(localStorage.getItem('timeRemaining'))}
            </div>
          </div>
          <MessageBoard
            location={this.props.location}
            timeRemaining={this.props.timeRemaining}
          />
        </div>
      </main>
    );
  }
}

PresentTimer.propTypes = {
  onStart: func.isRequired,
  onPause: func.isRequired,
  onClear: func.isRequired,
  handleDismiss: func.isRequired,
  timeRemaining: number,
  startClicked: bool,
  location: object
};

export default PresentTimer;
