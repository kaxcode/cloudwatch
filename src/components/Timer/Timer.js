import React from 'react';
import Alert from '../Alert/Alert';
import { oneOfType, string, func, number, bool, object } from 'prop-types';
import AdjustTimerButtonGroup from './AdjustTimerButtonGroup';
import MessageBoard from '../MessageBoard/MessageBoard.js';
import Nav from '../Nav/Nav.js';
import './Timer.css';
import ControllerButtons from '../ControllerButtons/ControllerButtons';
import TimeDisplay from '../TimeDisplay/TimeDisplay';

const Timer = props => {
  return (
    <main className="Timer">
      <Alert
        onDismiss={props.handleDismiss}
        show={props.startClicked && props.timeRemaining <= 0}
      />
      <Nav />
      <div className="Timer__Container">
        <AdjustTimerButtonGroup
          increaseHours={props.increaseHours}
          decreaseHours={props.decreaseHours}
          increaseMinutes={props.increaseMinutes}
          decreaseMinutes={props.decreaseMinutes}
          increaseSeconds={props.increaseSeconds}
          decreaseSeconds={props.decreaseSeconds}
        />
        <div className="Timer__Counter">
          <TimeDisplay time={props.timeRemaining} />
        </div>
        <ControllerButtons
          onPause={props.onPause}
          onStart={props.onStart}
          onClear={props.onClear}
        />
        <MessageBoard
          location={props.location}
          timeRemaining={props.timeRemaining}
        />
      </div>
    </main>
  );
};

Timer.propTypes = {
  onStart: func.isRequired,
  onPause: func.isRequired,
  onClear: func.isRequired,
  handleDismiss: func.isRequired,
  increaseHours: func.isRequired,
  decreaseHours: func.isRequired,
  increaseMinutes: func.isRequired,
  decreaseMinutes: func.isRequired,
  increaseSeconds: func.isRequired,
  decreaseSeconds: func.isRequired,
  timeRemaining: oneOfType([number, string]),
  startClicked: bool,
  location: object
};

export default Timer;
