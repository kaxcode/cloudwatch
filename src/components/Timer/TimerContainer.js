import React from 'react';
import Timer from './Timer.js';
import { func, object } from 'prop-types';

const HOURS = 3600 * 1000;
const MINUTES = 60 * 1000;
const SECONDS = 1 * 1000;
const MAX_TIME = 24 * 60 * 60 * 1000;
const MIN_TIME = 0;

class TimerContainer extends React.Component {
  static defaultProps = {
    now: Date.now
  };

  static propTypes = {
    location: object,
    now: func.isRequired
  };

  state = {
    lastTick: 0,
    timeRemaining: 0,
    startClicked: false
  };

  timer = null;

  componentDidMount = () => {
    const localStorageRef = localStorage.getItem('timeRemaining');
    this.setState({ timeRemaining: parseInt(localStorageRef, 0) || 0 });
  };

  componentDidUpdate() {
    localStorage.setItem('timeRemaining', this.state.timeRemaining);
  }

  componentWillUnmount() {
    localStorage.setItem('timeRemaining', 0);
  }

  increaseHours = () => {
    if (this.state.timeRemaining >= MAX_TIME) {
      return;
    }

    this.setState({ timeRemaining: this.state.timeRemaining + HOURS });
  };

  decreaseHours = () => {
    if (this.state.timeRemaining < HOURS) {
      return;
    }

    this.setState({ timeRemaining: this.state.timeRemaining - HOURS });
  };

  increaseMinutes = () => {
    if (this.state.timeRemaining >= MAX_TIME) {
      return;
    }

    this.setState({ timeRemaining: this.state.timeRemaining + MINUTES });
  };

  decreaseMinutes = () => {
    if (this.state.timeRemaining < MINUTES) {
      return;
    }

    this.setState({ timeRemaining: this.state.timeRemaining - MINUTES });
  };

  increaseSeconds = () => {
    if (this.state.timeRemaining >= MAX_TIME) {
      return;
    }

    this.setState({ timeRemaining: this.state.timeRemaining + SECONDS });
  };

  decreaseSeconds = () => {
    if (this.state.timeRemaining <= MIN_TIME) {
      return;
    }

    this.setState({ timeRemaining: this.state.timeRemaining - SECONDS });
  };

  handleStart = () => {
    if (this.state.startClicked) {
      return;
    }
    clearInterval(this.timer);
    this.setState({
      lastTick: this.props.now(),
      startClicked: true
    });
    this.timer = setInterval(this.tick, 10);
    localStorage.setItem('startClicked', true);
  };

  handlePause = () => {
    clearInterval(this.timer);
    this.setState({ startClicked: false });
    localStorage.setItem('startClicked', false);
  };

  handleClear = () => {
    this.setState({
      timeRemaining: 0
    });
    localStorage.setItem('timeRemaining', 0);
  };

  tick = () => {
    if (this.state.timeRemaining <= 0) {
      return this.handleClear();
    }

    const lastTick = this.props.now();
    const elapsedTime = lastTick - this.state.lastTick;
    this.setState({
      timeRemaining: this.state.timeRemaining - elapsedTime,
      lastTick
    });
  };

  handleDismiss = () => {
    this.setState({
      startClicked: false
    });
    clearInterval(this.timer);
  };

  render() {
    return (
      <Timer
        handleDismiss={this.handleDismiss}
        increaseHours={this.increaseHours}
        decreaseHours={this.decreaseHours}
        increaseMinutes={this.increaseMinutes}
        decreaseMinutes={this.decreaseMinutes}
        increaseSeconds={this.increaseSeconds}
        decreaseSeconds={this.decreaseSeconds}
        onStart={this.handleStart}
        onPause={this.handlePause}
        onClear={this.handleClear}
        startClicked={this.state.startClicked}
        timeRemaining={this.state.timeRemaining}
        location={this.props.location}
      />
    );
  }
}

TimerContainer.propTypes = {
  location: object
};

export default TimerContainer;
