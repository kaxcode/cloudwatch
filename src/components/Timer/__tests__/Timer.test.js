import React from 'react';
import Enzyme from 'enzyme';
import Timer from '../Timer';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });
jest.useFakeTimers();

describe('Timer', () => {
  let wrapper;

  const onStartMock = jest.fn();
  const onPauseMock = jest.fn();
  const onClearMock = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <Timer
        startClicked
        timeRemaining={0}
        onStart={onStartMock}
        onPause={onPauseMock}
        onClear={onClearMock}
      />
    );
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('shows alert when start is clicked and time is zero', () => {
    expect(wrapper.find('Alert').props().show).toBe(true);
  });

  it('does not show alert when start is not clicked and time is zero', () => {
    wrapper = shallow(<Timer startClicked={false} timeRemaining={0} />);
    expect(wrapper.find('Alert').props().show).toBe(false);
  });

  it('does not show alert when start is clicked and time is not zero', () => {
    wrapper = shallow(<Timer startClicked timeRemaining={10} />);
    expect(wrapper.find('Alert').props().show).toBe(false);
  });

  describe('the start button', () => {
    it('calls the :onStart prop when clicked', () => {
      //Arrange
      const startButton = wrapper.find('#timer-start').first();
      //Act
      startButton.simulate('click');
      //Assert
      expect(onStartMock).toHaveBeenCalled();
    });
  });

  describe('the pause button', () => {
    it('calls the :onPause prop when clicked', () => {
      //Arrange
      const pauseButton = wrapper.find('#timer-pause').first();
      //Act
      pauseButton.simulate('click');
      //Assert
      expect(onPauseMock).toHaveBeenCalled();
    });
  });

  describe('the clear button', () => {
    it('calls the :onClear prop when clicked', () => {
      //Arrange
      const clearButton = wrapper.find('#timer-clear').first();
      //Act
      clearButton.simulate('click');
      //Assert
      expect(onClearMock).toHaveBeenCalled();
    });
  });
});
