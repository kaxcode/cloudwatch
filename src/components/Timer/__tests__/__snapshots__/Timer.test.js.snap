// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer renders correctly 1`] = `
<main
  className="Timer"
>
  <Alert
    onDismiss={[Function]}
    show={true}
  />
  <Nav />
  <div
    className="Timer__Container"
  >
    <AdjustTimerButtonGroup
      decreaseHours={[Function]}
      decreaseMinutes={[Function]}
      decreaseSeconds={[Function]}
      increaseHours={[Function]}
      increaseMinutes={[Function]}
      increaseSeconds={[Function]}
    />
    <div
      className="Timer__Counter"
    >
      <TimeDisplay
        time={0}
      />
    </div>
    <ControllerButtons
      onClear={[Function]}
      onPause={[Function]}
      onStart={[Function]}
    />
    <MessageBoard
      showInput={true}
      timeRemaining={0}
      windowOpener={[Function]}
    />
  </div>
</main>
`;
